### PR TITLE
Bottom tabs navigation

### PR DESCRIPTION
**Pull request by vycius**
https://trello.com/c/D0427WhF/193-remove-main-screen-add-navigation-drawer
Introduced bottom navigation instead of homescreen. Bottom navigation uses BottomNavigationController. Also added new concept for navigation - NavigationManager with FragmentTransactionExecutor. Looking forward for comments and suggestions.

# TODO
- [x] Map filter doesn't work. Needs investigation.

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -103,6 +103,8 @@ dependencies {
     compile "com.android.support:recyclerview-v7:$supportLibraryVersion"
     compile "com.android.support:gridlayout-v7:$supportLibraryVersion"
     compile "com.android.support:design:$supportLibraryVersion"
+    compile "com.android.support:preference-v7:$supportLibraryVersion"
+    compile "com.android.support:preference-v14:$supportLibraryVersion"
```
Comment (mjurkus) [inline app/build.gradle]: Why 2 versions?

File: app/src/main/java/lt/vilnius/tvarkau/MainActivity.kt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * An activity representing a main activity home screen
+ */
+class MainActivity : BaseActivity() {
+
+    @Inject
+    lateinit var bottomNavigationController: BottomNavigationController
+
+    override fun onInject(applicationComponent: ApplicationComponent) {
+        MainActivityComponent.init(applicationComponent, this).inject(this)
```
Comment (mjurkus) [inline app/src/main/java/lt/vilnius/tvarkau/MainActivity.kt]: Why is this needed?

File: app/src/main/java/lt/vilnius/tvarkau/dagger/component/MainActivityComponent.kt [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    fun inject(myReportsListFragment: MyReportsListFragment)
+
+    fun inject(baseMapFragment: BaseMapFragment)
+
+    fun inject(fragment: MultipleProblemsMapFragment)
+
+
+    companion object {
+        fun init(applicationComponent: ApplicationComponent,
+                 activity: AppCompatActivity): MainActivityComponent {
```
Comment (mjurkus) [inline app/src/main/java/lt/vilnius/tvarkau/dagger/component/MainActivityComponent.kt]: when folding function arguments I would suggest it do it this way:
```
fun init(
    applicationComponent: ApplicationComponent,
    activity: AppCompatActivity
): MainActivityComponent
```

this way if new param is added or old removed only one line will be changed.

File: app/src/main/java/lt/vilnius/tvarkau/dagger/module/MainActivityModule.kt [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import lt.vilnius.tvarkau.navigation.NavigationManager
+
+
+@Module
+class MainActivityModule(private val activity: AppCompatActivity) {
+
+    @Provides
+    fun provideActivity(): Activity {
+        return activity
+    }
```
Comment (mjurkus) [inline app/src/main/java/lt/vilnius/tvarkau/dagger/module/MainActivityModule.kt]: inline

File: app/src/main/java/lt/vilnius/tvarkau/fragments/AllReportsListFragment.kt [hunk 5]
```diff
@@ -46,6 +52,7 @@ class AllReportsListFragment : BaseReportListFragment(), ReportListView {
         )
     }
 
+
```
Comment (mjurkus) [inline app/src/main/java/lt/vilnius/tvarkau/fragments/AllReportsListFragment.kt]: double spacing

File: app/src/main/java/lt/vilnius/tvarkau/fragments/BaseMapFragment.kt [hunk 6]
```diff
@@ -15,14 +16,22 @@ import com.google.android.gms.maps.CameraUpdateFactory
...
+import lt.vilnius.tvarkau.navigation.NavigationManager
 import lt.vilnius.tvarkau.views.adapters.MapsInfoWindowAdapter
+import javax.inject.Inject
 
 abstract class BaseMapFragment : BaseFragment(),
         GoogleMap.OnMarkerClickListener,
         GoogleApiClient.ConnectionCallbacks {
 
+    @Inject
+    lateinit var navigationManager: NavigationManager
```
Comment (mjurkus) [inline app/src/main/java/lt/vilnius/tvarkau/fragments/BaseMapFragment.kt]: Move to `BaseFragment`, keep code DRY

File: app/src/main/java/lt/vilnius/tvarkau/navigation/BottomNavigationController.kt [hunk 7]
```diff
@@ -0,0 +1,36 @@
+package lt.vilnius.tvarkau.navigation
+
+import android.app.Activity
+import android.support.annotation.IdRes
+import kotlinx.android.synthetic.main.activity_main.*
+import lt.vilnius.tvarkau.R
+
+
+class BottomNavigationController(private val activity: Activity, private val navigationManager: NavigationManager)
+    : NavigationController {
```
Comment (mjurkus) [inline app/src/main/java/lt/vilnius/tvarkau/navigation/BottomNavigationController.kt]:
```
class BottomNavigationController(
    private val activity: Activity, 
    private val navigationManager: NavigationManager
) : NavigationController 
```

File: app/src/main/java/lt/vilnius/tvarkau/navigation/BottomNavigationController.kt [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+    override fun onCreate(initial: Boolean) {
+        activity.bottom_navigation.setOnNavigationItemSelectedListener {
+            item ->
```
Comment (mjurkus) [inline app/src/main/java/lt/vilnius/tvarkau/navigation/BottomNavigationController.kt]: replace with `it`.

File: app/src/main/java/lt/vilnius/tvarkau/navigation/BottomNavigationController.kt [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    private fun onTabSelected(@IdRes id: Int) {
+        when (id) {
+            R.id.tab_list_of_problems -> navigationManager.navigateToMenuItem(NavigationManager.TabItem.REPORTS_LIST)
+            R.id.tab_map_of_problems -> navigationManager.navigateToMenuItem(NavigationManager.TabItem.REPORTS_MAP)
+            R.id.tab_my_problems -> navigationManager.navigateToMenuItem(NavigationManager.TabItem.MY_REPORTS_LIST)
+            R.id.tab_settings ->  {
+                navigationManager.navigateToMenuItem(NavigationManager.TabItem.SETTINGS)
+            }
```
Comment (mjurkus) [inline app/src/main/java/lt/vilnius/tvarkau/navigation/BottomNavigationController.kt]: Why braces for this one?

File: app/src/main/java/lt/vilnius/tvarkau/navigation/FragmentTransactionExecutor.kt [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        }
+    }
+
+    @SuppressLint("CommitTransaction")
+    fun replaceWithVerticalAnimation(fragment: Fragment, addToBackStack: Boolean) {
+        fragmentManager.beginTransaction().apply {
+            setCustomAnimations(R.anim.slide_in_from_top, 0, 0, R.anim.slide_out_to_top)
+            replace(R.id.container, fragment)
+
+
```
Comment (mjurkus) [inline app/src/main/java/lt/vilnius/tvarkau/navigation/FragmentTransactionExecutor.kt]: double spacing

File: app/src/main/java/lt/vilnius/tvarkau/navigation/FragmentTransactionExecutor.kt [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+class FragmentTransactionExecutor(private val fragmentManager: FragmentManager) {
+
+    fun replaceWithClearTop(fragment: Fragment) {
+        clearBackStackTop()
+        replace(fragment, false)
+    }
+
+    @SuppressLint("CommitTransaction")
+    fun replace(fragment: Fragment, addToBackStack: Boolean) {
+        fragmentManager.beginTransaction().apply {
```
Comment (mjurkus) [inline app/src/main/java/lt/vilnius/tvarkau/navigation/FragmentTransactionExecutor.kt]: don't use `apply`. It only introduces another nesting level and doesn't help with code readability.
For example this way looks better IMO:
```
fragmentManager.beginTransaction()
    .replace(R.id.container, fragment)
    .setAllowOptimization(true)
    .apply {
        if (addToBackStack) addToBackStack(null)
    }
    .commitAllowingStateLoss()
```

File: app/src/main/java/lt/vilnius/tvarkau/MainActivity.kt (hunk 2 above)
Comment (vycius) [reply app/src/main/java/lt/vilnius/tvarkau/MainActivity.kt]: To create MainActivityComponent and inject for instance BottomNavigationController

File: app/src/main/java/lt/vilnius/tvarkau/fragments/BaseMapFragment.kt (hunk 6 above)
Comment (vycius) [reply app/src/main/java/lt/vilnius/tvarkau/fragments/BaseMapFragment.kt]: I am still thinking about design. Currently we have MainActivity module with `NavigationManager` and `BottomNavigationController`. And `BottomNavigationController` should be injected only to MainActivity and it's fragment. `NavigationManager` should be injected to all activities and fragments. 
I would suggest to have `MainActivityModule` and `ActivityModule`. As it wouldn't let you inject `BottomNavigationController` into non MainActivity fragment.

File: app/src/main/java/lt/vilnius/tvarkau/MainActivity.kt (hunk 2 above)
Comment (mjurkus) [reply app/src/main/java/lt/vilnius/tvarkau/MainActivity.kt]: Don't duplicate same code
` MainActivityComponent.init(....).inject(this)` in all classes.

Create once in BaseActivity
```
    protected MainActivityComponent buildComponent(TvarkauApplication application) {
        return application.getComponent().mainActivityComponent(MainActivityModule(this));
    }
```

and pass via `onInject(component)`

Comment (vycius) [reply app/src/main/java/lt/vilnius/tvarkau/MainActivity.kt]: Yes. It will be done when I have some free time to finish. There is still a lot of to dos.

**Comment by vycius** [comment]
@mjurkus It's ready for review.

**Review by mjurkus — APPROVED** [review]
